content/artifacts/harvest-artifacts/plugin.js

```diff
@@ -0,0 +1,337 @@
+      .filter((p) => p.heldArtifactIds)
+      .filter((p) => p.heldArtifactIds.length > 0)
+      .filter((p) => p.planetLevel >= minPlanetLevel)
+      .filter((p) => !followedPlanetList.includes(p.locationId))
+      .sort((b, a) => a.planetLevel - b.planetLevel)
+      .slice(0, number)
+
+  planets.forEach((p) => {
+    let arts = p.heldArtifactIds;
+    console.log(`working on L${p.planetLevel} ${Object.keys(planetTypes)[p.planetType]} [${arts.length}]ARTs. ${p.locationId}`);
+
+    for (let i = 0; i < arts.length; ++i) {
+      if (!df.getArtifactWithId(arts[i])) {
+        ///HACK
+        //          if (df.getArtifactWithId(arts[i].artifactType <=4) )  //only get consumables
+        //         {
+        //          console.log ("skipping non consumables"); 
+        //          return;
+        //         }  
+        console.log(" --artifact not found:", arts[i]);
+        break;
+      }
+
+      if (isArtifactActivated(df.getArtifactWithId(arts[i]))) {  //TODO: check for time
+        df.terminal.current.println(` --artifact is activated; ${autoDeactivate ? "deactivating  now" : "Not deactivating"}`);
+        console.log(` --artifact is activated; ${autoDeactivate ? "deactivating  now" : "Not deactivating"}`);
+        if (!reviewonly && autoDeactivate) df.deactivateArtifact(p.locationId, arts[i]);
+        break;
+      }
+
+      if ((p.planetType == planetTypes["Spacetime Rip"]) && (p.planetLevel > (df.getArtifactWithId(arts[i]).rarity))  // art at rip
+      ) {
+        console.log(" -- withdrawing", p.locationId, arts[i], df.getArtifactWithId(arts[i]).rarity);
+        df.terminal.current.println(`...withdrawing from ${p.locationId}`);
+
+        if (!reviewonly) df.withdrawArtifact(p.locationId, arts[i]);
+        continue;
+      }
+      //do transport
+      if (!reviewonly || 1 == 1) {  //<<<<<<<<<
+        transportART2(p.locationId, arts[i], 98, "", errorPlanets);  //hard code to 98% of range, for peace time
+        return; // this should set it to only transport 1 single art per planet; additional arts will be dealt on next run
+      }
+    }
+  })
+
+
+  return errorPlanets;
+}
+
+function transportART2(srcId, artId, maxRangePct = 50, tgtId = "", errorPlanets = []) {
+  const myART = df.getArtifactWithId(artId);
+  if (!myART) { console.log(" --invalid ART"); return }
+
+  const source = df.getPlanetWithId(srcId);
+
+  if (source.owner != df.account) { console.log(" --Dont Own planet"); return }
+
+  //#1, try send to a RIP
+  if (tgtId === "") {
+    let targetList = df.getMyPlanets()
+      .filter((p) => df.getLocationOfPlanet(p.locationId))
+      .filter((p) => p.planetType == 3)  //only to RIP
+      .filter((p) => p.planetLevel > myART.rarity)
+      .filter((p) => (df.getDist(srcId, p.locationId) < df.getMaxMoveDist(srcId, maxRangePct)))
+      .filter((p) => p.locationId !== srcId)
+      .sort((a, b) => { return df.getDist(srcId, a.locationId) - df.getDist(srcId, b.locationId) })
```

Comment:
I see a lot of these .filter() chains
this could be rewritten to only use 1 .filter() and all conditions could be inside of it separated with a &&
I am not a js expert so please correct me if I am wrong, but i think every time the filter function creates a new array of planets and this requires more time.
Not saying that this needs to be changed, maybe there is no performance problem here.

Reply:
wow. Difference is quite clear.  Thanks.  Learned something. This will be useful to know for bots.   For this code though, there is no performance issue.